File: frontend/src/api/knowledge.ts
import http from './http'
import type { ApiEnvelope } from '@/types/api'

export interface KnowledgeIndexStatus {
  provider: 'local_keyword' | 'chroma' | string
  indexed: boolean
  document_count: number
  clause_count: number
  directory: string | null
  collection: string | null
}

export interface KnowledgeSearchResult {
  id: string
  document_id: string
  title: string
  source: string
  article: string
  version: string
  category: string
  effective_date: string | null
  content: string
  metadata: Record<string, unknown>
  status?: string
  created_at?: string
  score?: number
}

export type KnowledgeDocument = KnowledgeSearchResult

export interface KnowledgeChatPayload {
  question: string
  project_id?: string | null
  use_llm?: boolean | null
}

export interface KnowledgeChatCitation {
  type: string
  document_id: string
  source: string
  article: string
  title: string
  score: number
}

export interface KnowledgeChatResult {
  question: string
  mode: 'rag_only' | 'rag_llm' | string
  description: string
  answer: string
  citations: KnowledgeChatCitation[]
  retrieval: {
    clauses: { ready: boolean; count: number }
    risk_tip: { included: boolean; hazard_types: string[] }
    site: { included: boolean; project_id: string | null }
  }
  llm: { used: boolean; model: string | null; error: string | null }
}

export const knowledgeApi = {
  async search(query = ''): Promise<KnowledgeSearchResult[]> {
    const response = await http.get<ApiEnvelope<KnowledgeSearchResult[]>>('/knowledge/search', { params: { q: query } })
    return response.data.data
  },
  async indexStatus(): Promise<KnowledgeIndexStatus> {
    const response = await http.get<ApiEnvelope<KnowledgeIndexStatus>>('/knowledge/index/status')
    return response.data.data
  },
  async reindex(): Promise<KnowledgeIndexStatus> {
    const response = await http.post<ApiEnvelope<KnowledgeIndexStatus>>('/knowledge/reindex')
    return response.data.data
  },
  async chat(payload: KnowledgeChatPayload): Promise<KnowledgeChatResult> {
    const response = await http.post<ApiEnvelope<KnowledgeChatResult>>('/knowledge/chat', payload)
    return response.data.data
  },
}

File: backend/app/schemas/knowledge.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class KnowledgeDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    document_id: str
    title: str
    source: str
    article: str
    version: str
    category: str
    effective_date: str | None
    content: str
    metadata: dict[str, object]
    status: str
    created_at: datetime


class KnowledgeDocumentCreate(BaseModel):
    title: str
    source: str
    version: str = "MVP"
    article: str = ""
    category: str
    effective_date: str | None = None
    content: str
    metadata: dict[str, object] = Field(default_factory=dict)


class KnowledgeSearchResult(BaseModel):
    id: str = ""
    document_id: str
    title: str = ""
    source: str
    article: str
    category: str = ""
    content: str
    version: str = ""
    effective_date: str | None = None
    score: float
    metadata: dict[str, object] = Field(default_factory=dict)


class KnowledgeChatRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)
    project_id: str | None = None
    use_llm: bool | None = None


class KnowledgeChatResult(BaseModel):
    question: str
    mode: str  # rag_only | rag_llm
    description: str
    answer: str
    citations: list[dict[str, object]]
    retrieval: dict[str, object]
    llm: dict[str, object]
